### PR TITLE
Add support for arbitrary input headers

### PR DESCRIPTION
**Pull request by desimmons**
## What does this PR do

* Adds support for arbitrary header input

## How should this PR be tested?

* Follow the "link API journey" to the header setting dialog
* Add a URL
* Add a new header
* remove __all__ the headers
* Make sure Next not available
* Add a header
* Input Key into input box and assert "Headers must be complete" box is still there
* Input Value into input box and assert that request headers are notified
* Add another input and assert the "Headers must be complete" box is visible again
* Remove empty header and progress

Chaos engineering:
* Try getting past without inputting all into all available input boxes
* Find any edge cases where the notified headers is visible when any header input is blank

## Checklist before requesting a review
- [ ] I have performed a self-review of my code

### CODE REVIEW

**Comment by robbie-daniels** [comment]
* Manually tested, couldn't find a way to break the headers
* Seems to work as expected

File: integration_test/linking_integration_test.dart [hunk 1]
```diff
@@ -6,8 +6,12 @@ import 'package:uuid/uuid.dart';
 
 var uuid = const Uuid();
 String theInputAssetName = "integration-test-api";
-const String theAPIURL = "https://api.xrpldata.com/api/v1";
-const String theAPIKey = "integration-test-key";
+// The whitespace should not be entered
+const String theAPIURL = "https://api.xrplda ta.com/api/v1";
```
Comment (robbie-daniels) [inline integration_test/linking_integration_test.dart:10]: Why's there a space in here?

Comment (desimmons) [reply integration_test/linking_integration_test.dart:10]: It's just an erroneous URL string. When we enter it, we assert that the whitespace is not also entered

Comment (robbie-daniels) [reply integration_test/linking_integration_test.dart:10]: I realised this after deleting the comment, sorry I thought I'd deleted it!

File: lib/marketplace/marketplace_dialogs.dart [hunk 2]
```diff
@@ -1294,8 +1294,152 @@ class _ChargeWidgetState extends State<ChargeWidget> {
...
+  Widget keyValueWidget(int index) {
+    return Row(
+      mainAxisSize: MainAxisSize.min,
+      children: [
+        Flexible(
+            child: TextField(
+          key: Key('Key ${index + 1}'),
+          inputFormatters: [
+            FilteringTextInputFormatter.allow(
+                RegExp(r'[^\s]+')), // This regex disallows any whitespace
```
Comment (robbie-daniels) [inline lib/marketplace/marketplace_dialogs.dart:1355]: I had a look to see what characters are valid for a request header, I think the answer is maybe not straightforward: https://en.wikipedia.org/wiki/Percent-encoding

Did you read somewhere that only whitespace needs to be excluded?

File: lib/marketplace/marketplace_dialogs.dart [hunk 3]
```diff
@@ -1310,84 +1454,128 @@ class LinkedAPIDetailsWidget extends StatefulWidget {
...
+        String value = apiKeyValueControllers[i].text;
+        String shortenedValue =
+            value.substring(0, min(maxCharsOfKey, value.length));
+        String line =
+            "'$key: ${value.length > maxCharsOfKey ? "$shortenedValue..." : value}'";
+        lines.add(line);
+      }
+    }
+    if (allHeadersAreComplete()) {
+      return "Requests will have header \n\n${lines.join("\n")}";
```
Comment (robbie-daniels) [inline lib/marketplace/marketplace_dialogs.dart:1497]: Nit: is it "headers"?

**Review by robbie-daniels — COMMENTED** [review]
* This doesn't seem to scale particularly well for mobile.  Do we plan to support mobile journeys for API linking?  I think it's a more realistic scenario than Dhali hosted, but I don't really have a feel for this at the moment.  Should we just disable mobile journeys for now?mobile-dims

* I think a scrollbar for when we've got a lot of headers would be good

* Could we reverse the order of the plus and minus buttons?

* The buttons and input boxes feel a bit "close" together - did you try other spacing too?

File: lib/marketplace/marketplace_dialogs.dart (hunk 2 above)
Comment (desimmons) [reply lib/marketplace/marketplace_dialogs.dart:1355]: I read that whitespace was OK for the value, but not the key

**Comment by robbie-daniels** [comment]
> This doesn't seem to scale particularly well for mobile. Do we plan to support mobile journeys for API linking? I think it's a more realistic scenario than Dhali hosted, but I don't really have a feel for this at the moment. Should we just disable mobile journeys for now?

Tested this by hard refreshing in mobile mode, and actually it looks pretty good - I retract this comment.

File: lib/marketplace/marketplace_dialogs.dart (hunk 2 above)
Comment (robbie-daniels) [reply lib/marketplace/marketplace_dialogs.dart:1355]: I'm thinking other characters than just whitespace - is your understanding that they are all fine?  I'm not sure, reading the page I linked above.

Comment (desimmons) [reply lib/marketplace/marketplace_dialogs.dart:1359]: No no, I'm not claiming that just white spaces are disallowed

**Pull request merged**